# <div class = "alert alert-info"> Data Cleaning

    Data cleaning means fixing bad data in your data set.Bad data could be:
- Empty cells
- Data in wrong format
- Wrong data
- Duplicates
    
We will learn how to deal with all of them.

In [1]:
import pandas as pd

In [2]:
path = "./data_raw.csv"
data_frame = pd.read_csv(path)

### <div class = "alert alert-success"> Our Data Set :
    
This dataset represents the workout session of a person. it has five attributes/columns and 32 records/rows (0-31)

In [3]:
print(data_frame.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45

<div class = "alert alert-success">
    
<div class = "alert alert-info">
    
- The data set contains some empty cells ("Date" in row 22, and "Calories" in row 18 and 28).
- The data set contains wrong format ("Date" in row 26).
- The data set contains wrong data ("Duration" in row 7).
- The data set contains duplicates (row 11 and 12).

## <div class = "alert alert-success"> (1) Cleaning Missing Values/Empty Cells
    
Empty cells can potentially give you a wrong result when you analyze data. To remove empty cells we can
- Remove rows with empty cells (usually done when dataset is large so that impact of deleting row is not significant)
- Replacing rows with empty cells



###  - <ins>Removing rows with missing data</ins>:

To remove row with missing values we use <span style="color:red">
dropna()
    </span> method. 
- dropna() method returns a new DataFrame, and will not change the original.
- dropna( inplace = True) method remove missing value from the orignal dataset and return nothing

In [4]:
data_frame_new = data_frame.dropna()  # missing values removed from copy of data_frame and stored in data_frame_new

print(data_frame_new.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
19        60

- row 18,22 and 28 that had missing values are now removed from the dataset

In [5]:
data_frame.dropna(inplace = True) # dropna(inplace= True) deteted row 18,22 & 28 from orignal data_frame and return nothing 
print(data_frame.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
19        60

### - <ins>Replacing rows with missing data</ins>: 

Two ways to replace missing values:
- Insert some value instead of missing value.The <span style="color:red">
fillna( )
    </span> method allows us to replace missing value/empty cell/NaN/Null with some value:
- Replace missing value with the mode, median or mean of the column from which its missing

 #####  <span style="color:green"> Replacing missing value by some value: </span> 

In [6]:
data_frame = pd.read_csv(path)   # loading again because we removed missing values rows above

data_frame.fillna(9999, inplace = True) # replace all mising value cells with 9999

print(data_frame.to_string())  

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45

- row 18, 22 & 28 missing values replaced by 9999

In [7]:
data_frame = pd.read_csv(path)    

data_frame["Calories"].fillna(9999, inplace = True) # replace Calories attribute mising value cells with 9999 only
data_frame["Date"].fillna('0/0/0', inplace = True) # replace Calories attribute mising value cells with 0/0/0

print(data_frame.to_string()) 

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45

 #####  <span style="color:green"> Replacing missing value by mean, median or mod: </span> 
 
 Pandas uses the <span style="color:red">
 mean(), median() and mode()
    </span> methods to calculate the respective values for a specified column:

 ######  <ins><span style="color:blue"> Replacing by mean<ins>:</span> 

In [8]:
data_frame = pd.read_csv('data_raw.csv')   # loading again because we changed the missing values above

data_frame_mean = data_frame["Calories"].mean()

print(f"Calories column mean is: {data_frame_mean}")

data_frame["Calories"].fillna(data_frame_mean, inplace = True)

print(data_frame.to_string())

Calories column mean is: 304.68
    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130    409.10
1         60   12/2/2020    117       145    479.00
2         60   12/3/2020    103       135    340.00
3         45   12/4/2020    109       175    282.40
4         45   12/5/2020    117       148    406.00
5         60   12/6/2020    102       127    300.00
6         60   12/7/2020    110       136    374.00
7        450   12/8/2020    104       134    253.30
8         30   12/9/2020    109       133    195.10
9         60  12/10/2020     98       124    269.00
10        60  12/11/2020    103       147    329.30
11        60  12/12/2020    100       120    250.70
12        60  12/12/2020    100       120    250.70
13        60  12/13/2020    106       128    345.30
14        60  12/14/2020    104       132    379.30
15        60  12/15/2020     98       123    275.00
16        60  12/16/2020     98       120    215.20
17        60  12/17/2020    100 

- row 18 and 28 of Calories column replaced by mean = 304.68

 ######  <ins><span style="color:blue"> Replacing by median<ins>:</span> 

In [9]:
data_frame = pd.read_csv('data_raw.csv')   # loading again because we changed the missing values above

data_frame_median = data_frame["Calories"].median()

print(f"Calories column median is: {data_frame_median}")

data_frame["Calories"].fillna(data_frame_median, inplace = True)

print(data_frame.to_string())

Calories column median is: 291.2
    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100

- row 18 and 28 of Calories column replaced by median = 291.2

 ######  <ins><span style="color:blue"> Replacing by mode<ins>: <span> 

In [10]:
data_frame = pd.read_csv('data_raw.csv')   # loading again because we changed the missing values above

data_frame_mode = data_frame["Calories"].mode()[0]  # without [0] Calories NaN are not replaced with mode= 300.0

print(f"Calories column mode is: {data_frame_mode}")

data_frame["Calories"].fillna(data_frame_mode, inplace = True)

print(data_frame.to_string())

Calories column mode is: 300.0
    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7        450   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100  

- row 18 and 28 of Calories column replaced by mode = 300.0

## <div class = "alert alert-success"> (2) Cleaning Data of Wrong Format
    
Cells with data of wrong format can make it hard to analyze data, let alone deduce meaningful insights.
Wrong fornated data can be cleaned in two waya:
- Remove the rows
- Convert all cells in the columns into the same format
    
In our Data Frame, we have two cells with the wrong format:row 22 and 26 of the 'Date' column. It shouldn't be an empty cell or integer value but rather a string that represents a date.
    
To convert all cells in the 'Date' column into dates we can use Pandas 
<span style="color:red">
to_datetime()
    </span>
method:

In [11]:
data_frame['Date'] = pd.to_datetime(data_frame['Date'])

print(data_frame.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

- The date in row 26 is fixed, but the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value. One way to deal with empty values is simply removing the entire row:

In [12]:
data_frame.dropna(subset=['Date'], inplace = True) # removing row 22
print(data_frame.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

## <div class = "alert alert-success"> (3) Fixing Wrong Data
  
Wrong data does not have to be "empty cells" or "wrong format", it can just be wrong. For example a typo like registering "5.7Kg" instead of "57Kg" for a fully grown person.
Sometimes one can spot wrong data just by looking at the data set other times it can go unchecked.

In our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.
It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not trully workout in 450 minutes (7.5 hrs!).
 
To fix wrong data we can:
- Replace the wrong values by correct one
- Remove rows with wrong values

 ######  <ins><span style="color:blue"> Replacing wrong values<ins>: <span> 

In [13]:
data_frame.loc[7, 'Duration'] = 49     # repllacing cell value in (row,column)= (7,Duration) with 33
print(data_frame.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         49 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

- For small data sets above method might be able to replace the wrong data one by one, but not for big data sets.
- To replace wrong data for larger data sets we can create some rules, say we set some boundaries for legal values, and replace any values that are outside of that boundaries. Below code explains this concept:

In [21]:
data_frame = pd.read_csv('data_raw.csv')
for data_frame_max_range in data_frame.index:
  if data_frame.loc[data_frame_max_range, "Duration"] > 60:  
  
    data_frame.loc[data_frame_max_range, "Duration"] = 60

print(data_frame.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
7         60   12/8/2020    104       134     253.3
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45

- (row 7,Duration) cell replaces 60 by 60
- ignore the NaN and wrong format data as we uploaded the data_raw.csv file anew

 ######  <ins><span style="color:blue"> Removing rows with wrong values<ins>: <span> 

 #####  <span style="color:green"> (i) From small datasets: </span> 

In [23]:
data_frame = pd.read_csv('data_raw.csv')

data_frame.drop(7, inplace = True)     #every cell in Duration column grater then or equal to 50 will be deleted

print(data_frame.to_string())   

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45  12/18/2020     90       112       NaN
19        60

#####  <span style="color:green"> (ii) From large datasets: </span> 

In [22]:
data_frame = pd.read_csv('data_raw.csv')
for data_frame_cell_value in data_frame.index:
      if data_frame.loc[data_frame_cell_value, "Duration"] > 60:
        data_frame.drop(data_frame_cell_value, inplace = True)  
        #every cell in Duration column grater then or equal to 50 will be deleted

print(data_frame.to_string())   

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
12        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45  12/18/2020     90       112       NaN
19        60

## <div class = "alert alert-success"> (4) Removing Duplicates
    
rowes in a dataframe registered more then once are considered as duplicates. In our test data set, note that row 11 and 12 are duplicates.

To discover duplicates in large datasets, we can use the <span style="color:red">
duplicated()
    </span> method.

- duplicated() method returns a Boolean values for each row- returns True for every row that is a duplicate, othwerwise False:

In [25]:
print(data_frame.duplicated())   # shows TRue against row 12 indacating its duplicate of row above (i.e row 11)

0     False
1     False
2     False
3     False
4     False
5     False
6     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


- To remove duplicates, use the <span style="color:red">
drop_duplicates() 
    </span>method.

In [27]:
data_frame.drop_duplicates(inplace = True)

print(data_frame.to_string())  # row 12 deleted

    Duration        Date  Pulse  Maxpulse  Calories
0         60   12/1/2020    110       130     409.1
1         60   12/2/2020    117       145     479.0
2         60   12/3/2020    103       135     340.0
3         45   12/4/2020    109       175     282.4
4         45   12/5/2020    117       148     406.0
5         60   12/6/2020    102       127     300.0
6         60   12/7/2020    110       136     374.0
8         30   12/9/2020    109       133     195.1
9         60  12/10/2020     98       124     269.0
10        60  12/11/2020    103       147     329.3
11        60  12/12/2020    100       120     250.7
13        60  12/13/2020    106       128     345.3
14        60  12/14/2020    104       132     379.3
15        60  12/15/2020     98       123     275.0
16        60  12/16/2020     98       120     215.2
17        60  12/17/2020    100       120     300.0
18        45  12/18/2020     90       112       NaN
19        60  12/19/2020    103       123     323.0
20        45